File: java/src/tarehart/rlbot/steps/rotation/PitchToPlaneStep.java
package tarehart.rlbot.steps.rotation;

import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.vector.Vector3;

public class PitchToPlaneStep extends OrientToPlaneStep {

    public PitchToPlaneStep(Vector3 planeNormal) {
        super(planeNormal);
    }

    public PitchToPlaneStep(Vector3 planeNormal, boolean allowUpsideDown) {
        super(planeNormal, allowUpsideDown);
    }

    @Override
    protected double getOrientationCorrection(CarData car) {
        Vector3 vectorNeedingCorrection = car.getOrientation().getNoseVector();
        Vector3 axisOfRotation = car.getOrientation().getRightVector();
        double correction = getMinimalCorrectionRadiansToPlane(vectorNeedingCorrection, axisOfRotation);

        boolean upsideDown = car.getOrientation().getRoofVector().dotProduct(planeNormal) < 0;

        if (upsideDown) {
            correction *= -1; // When upside down, need to rotate the opposite direction to converge on plane.
            if (!allowUpsideDown) {
                correction += Math.PI; // Turn all the way around
            }
        }
        return RotationUtil.shortWay(correction);
    }

    @Override
    protected double getAngularVelocity(CarData car) {
        return car.getSpin().getPitchRate();
    }

    @Override
    protected AgentOutput accelerate(boolean positiveRadians) {
        return  new AgentOutput().withPitch(positiveRadians ? 1 : -1);
    }

    @Override
    protected double getSpinDeceleration() {
        return 6;
    }


    @Override
    public String getSituation() {
        return "Pitching in midair";
    }

}

File: src/main/java/tarehart/rlbot/steps/rotation/PitchToPlaneStep.kt
package tarehart.rlbot.steps.rotation

import tarehart.rlbot.AgentOutput
import tarehart.rlbot.input.CarData
import tarehart.rlbot.math.vector.Vector3

class PitchToPlaneStep(planeNormal: Vector3, allowUpsideDown: Boolean = false) : OrientToPlaneStep({planeNormal}, allowUpsideDown) {

    override val spinDeceleration: Double = 6.0

    override val situation = "Pitching in midair"


    override fun getOrientationCorrection(car: CarData): Double {
        val vectorNeedingCorrection = car.orientation.noseVector
        val axisOfRotation = car.orientation.rightVector
        var correction = getMinimalCorrectionRadiansToPlane(vectorNeedingCorrection, axisOfRotation)

        val upsideDown = car.orientation.roofVector.dotProduct(planeNormal) < 0

        if (upsideDown) {
            correction *= -1.0 // When upside down, need to rotate the opposite direction to converge on plane.
            if (!allowUpsideDown) {
                correction += Math.PI // Turn all the way around
            }
        }
        return RotationUtil.shortWay(correction)
    }

    override fun getAngularVelocity(car: CarData): Double {
        return car.spin.pitchRate
    }

    override fun accelerate(positiveRadians: Boolean): AgentOutput {
        return AgentOutput().withPitch((if (positiveRadians) 1 else -1).toDouble())
    }

}
